File: regions/system/library/dart/encryption/lib/src/aes.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:encrypt/encrypt.dart' as enc;

/// Generates a cryptographically secure random 256-bit (32-byte) AES key.
Uint8List generateKey() {
  final rng = Random.secure();
  return Uint8List.fromList(List.generate(32, (_) => rng.nextInt(256)));
}

/// Encrypts [plaintext] using AES-GCM with the given 256-bit [key].
///
/// A random 12-byte IV/nonce is generated for each call. The returned string
/// is a base64-encoded concatenation of:
///   IV (12 bytes) + ciphertext + GCM auth tag (16 bytes)
///
/// This format is self-contained: the [decrypt] function extracts the IV and
/// auth tag automatically.
String encrypt(Uint8List key, String plaintext) {
  final encrypter = enc.Encrypter(enc.AES(enc.Key(key), mode: enc.AESMode.gcm));
  final iv = enc.IV.fromSecureRandom(12);
  final encrypted = encrypter.encrypt(plaintext, iv: iv);

  // Concatenate: IV (12 bytes) + ciphertext + auth tag (16 bytes)
  final combined = Uint8List.fromList(iv.bytes + encrypted.bytes);
  return base64.encode(combined);
}

/// Decrypts [ciphertext] that was produced by [encrypt] using the same [key].
///
/// Extracts the 12-byte IV from the front and the 16-byte GCM auth tag from
/// the end, then decrypts and authenticates the middle portion.
///
/// Throws if the auth tag is invalid (i.e. the data was tampered with or the
/// wrong key was used).
String decrypt(Uint8List key, String ciphertext) {
  final combined = base64.decode(ciphertext);

  // Extract IV (first 12 bytes)
  final iv = enc.IV(Uint8List.fromList(combined.sublist(0, 12)));

  // The rest is ciphertext + GCM auth tag (handled by the encrypt package)
  final encryptedBytes = Uint8List.fromList(combined.sublist(12));
  final encrypted = enc.Encrypted(encryptedBytes);

  final encrypter = enc.Encrypter(enc.AES(enc.Key(key), mode: enc.AESMode.gcm));
  return encrypter.decrypt(encrypted, iv: iv);
}

File: regions/system/library/python/encryption/src/k1s0_encryption/aes.py
"""AES-GCM authenticated encryption.

Uses the ``cryptography`` library's AESGCM primitive with a random
12-byte nonce per encryption.  The wire format is::

    base64( nonce ‖ ciphertext ‖ tag )

where *nonce* is 12 bytes, and the 16-byte authentication tag is
appended by AESGCM automatically.
"""

from __future__ import annotations

import base64
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

_NONCE_SIZE = 12  # 96-bit nonce recommended by NIST for AES-GCM


def generate_key() -> bytes:
    """Generate a random 256-bit (32-byte) AES key."""
    return AESGCM.generate_key(bit_length=256)


def encrypt(key: bytes, plaintext: str) -> str:
    """Encrypt *plaintext* with AES-256-GCM and return a base64 string.

    A fresh random nonce is generated for every call so identical
    plaintexts produce different ciphertexts.

    Parameters
    ----------
    key:
        A 32-byte AES key (use :func:`generate_key`).
    plaintext:
        UTF-8 text to encrypt.

    Returns
    -------
    str
        Base64-encoded ``nonce + ciphertext + tag``.
    """
    nonce = os.urandom(_NONCE_SIZE)
    aesgcm = AESGCM(key)
    ct = aesgcm.encrypt(nonce, plaintext.encode("utf-8"), None)
    return base64.b64encode(nonce + ct).decode("ascii")


def decrypt(key: bytes, ciphertext: str) -> str:
    """Decrypt a base64 AES-GCM ciphertext produced by :func:`encrypt`.

    Parameters
    ----------
    key:
        The same 32-byte AES key used for encryption.
    ciphertext:
        Base64 string returned by :func:`encrypt`.

    Returns
    -------
    str
        The original plaintext.

    Raises
    ------
    cryptography.exceptions.InvalidTag
        If the key is wrong or the data has been tampered with.
    """
    raw = base64.b64decode(ciphertext)
    nonce = raw[:_NONCE_SIZE]
    ct = raw[_NONCE_SIZE:]
    aesgcm = AESGCM(key)
    return aesgcm.decrypt(nonce, ct, None).decode("utf-8")
